// @flashbots/mev-share-client/build/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedNetworks = exports.mungeSendBundleParams = exports.mungePrivateTxParams = exports.StreamEvent = void 0;
var interfaces_1 = require("./interfaces");
Object.defineProperty(exports, "StreamEvent", { enumerable: true, get: function () { return interfaces_1.StreamEvent; } });
var mungers_1 = require("./mungers");
Object.defineProperty(exports, "mungePrivateTxParams", { enumerable: true, get: function () { return mungers_1.mungePrivateTxParams; } });
Object.defineProperty(exports, "mungeSendBundleParams", { enumerable: true, get: function () { return mungers_1.mungeSendBundleParams; } });
/**
 * Network connection presets supported by the Flashbots Matchmaker.
 */
exports.supportedNetworks = {
    mainnet: {
        name: "mainnet",
        chainId: 1,
        streamUrl: "https://mev-share.flashbots.net",
        apiUrl: "https://relay.flashbots.net"
    },
    goerli: {
        name: "goerli",
        chainId: 5,
        streamUrl: "https://mev-share-goerli.flashbots.net",
        apiUrl: "https://relay-goerli.flashbots.net"
    },
};
