// @flashbots/mev-share-client/build/api/networks.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupportedNetworks = void 0;
const error_1 = require("../error");
/**
 * Network connection presets supported by Flashbots.
 */
class SupportedNetworks {
    /**
     * Returns true if the given chainId is supported by the client.
     */
    static supportsChainId(chainId) {
        return Object.values(this.supportedNetworks).map(n => n.chainId).includes(chainId);
    }
    /** Gets the network preset matching the provided chainId, throws an error if not found. */
    static getNetwork(chainId) {
        if (this.supportsChainId(chainId)) {
            const net = Object.values(this.supportedNetworks).find(net => net.chainId === chainId);
            if (net) {
                return net;
            }
        }
        throw new error_1.UnimplementedNetwork({ chainId });
    }
}
exports.SupportedNetworks = SupportedNetworks;
_a = SupportedNetworks;
SupportedNetworks.supportedNetworks = {
    mainnet: {
        name: "mainnet",
        chainId: 1,
        streamUrl: "https://mev-share.flashbots.net",
        apiUrl: "https://relay.flashbots.net"
    },
    goerli: {
        name: "goerli",
        chainId: 5,
        streamUrl: "https://mev-share-goerli.flashbots.net",
        apiUrl: "https://relay-goerli.flashbots.net"
    }
};
// expose networks individually as class properties
SupportedNetworks.mainnet = _a.supportedNetworks.mainnet;
SupportedNetworks.goerli = _a.supportedNetworks.goerli;
